# Task 4

In [0]:
# Loading modules that we need
from pyspark.sql.session import SparkSession
from pyspark.sql.types import *
from pyspark.sql.functions import *
from typing import Any
from pyspark.sql.functions import corr
from pyspark.sql import functions as F


#### Connection to Local spark Cluster

In [ ]:
# Connect to the Spark Master running in Docker
spark = SparkSession.builder \
    .appName("MyLocalNotebook4") \
    .master("spark://localhost:7077") \
    .config("spark.driver.host", "localhost") \
    .config("spark.driver.memory", "1g") \
    .config("spark.executor.memory", "1g") \
    .getOrCreate()

# Verify connection
print(spark.version)
print("Spark is running on", spark.sparkContext.master)

In [0]:
# A helper function to load a table (stored in Parquet format) from DBFS as a Spark DataFrame 
def load_df(table_name: "name of the table to load") -> DataFrame:
    return spark.read.format("delta").load(table_name)

users_df = load_df("users")
posts_df = load_df("posts")

# Uncomment if you need
# comments_df = load_df("comments")
# badges_df = load_df("badges")

#### The problem: mining the interests of experts

The primary role of a questions and answering platform such as Stack Exchange is to connect two types of people. Namely, people who have questions in areas such as computer science or data science and knowledgeable people who can answer those questions reliably. Let's call the first category of people' knowledge seekers' and the second one 'expert users' or 'experts' for short.

Here we want to answer a question related to the diversity of topics that experts are interested in using our data. We want to know if expert users only answer questions in a specific set of topics or their interests include a wide variety of topics.

To answer the above question, we will compute the correlation between a user's expertise level and the diversity of topics of questions they have answered. The first step is to define two variables (or measures); first for 'user expertise level' and then for 'user interest diversity'. Then we will use the Pearson correlation coefficient to measure the linear correlation between the two variables. We define the variables as:

   - VariableA (the measure of user expertise level). We will use the 'Reputation' column from 'users' table, which according to Stack Exchange's documentation "is a rough measurement of how much the community trusts you; it is earned by convincing your peers that you know what you're talking about" as an indicator of a user's expertise level on the platform. 

   - VariableB (The measure of user interest diversity). We measure the diversity of a user's interests by computing the total number of distinct tags associated with the questions each user has answered divided by the total number of unique tags which is 638.

Compute the Pearson correlation coefficient between VariableA and VariableB, and based on the result you've got, answer the following question: 

     Do expert users have specif interests or do they have general interests?

Please explain your reasoning on how you reached your answer.

You should use Apache Spark API for your implementation. You can use the Spark implementation of the Pearson correlation coefficient.

In [0]:

df_varA = users_df.select("Id", "Reputation")
df_varA.show()





+---+----------+
| Id|Reputation|
+---+----------+
| -1|         1|
|  1|       101|
|  2|       101|
|  3|       101|
|  4|       101|
|  5|       215|
|  6|       101|
|  7|       101|
|  8|       101|
|  9|      1102|
| 10|       101|
| 11|       213|
| 12|       101|
| 14|      2782|
| 15|       101|
| 16|         1|
| 17|       236|
| 18|       101|
| 19|       101|
| 20|       101|
+---+----------+
only showing top 20 rows



In [0]:

# Find the number of distinct tags associated with a question each user has answered
# divide this number by fetch number of unique tags
def run_query2(query: "a SQL query string", df1: "DataFrame A", df2: "DataFrame B") -> Any:
    df1.createOrReplaceTempView("df1")
    df2.createOrReplaceTempView("df2")
    result = spark.sql(query)
    return result



q1 = "SELECT df2.Id, df1.Tags, df2.OwnerUserId FROM df1 INNER JOIN df2 ON df1.Id = df2.ParentId GROUP BY df1.Tags, df2.Id, df2.OwnerUserId ORDER BY df2.Id ASC"
unique_nb_tags = 638

count_splits = udf(lambda x : len(x) - 1, IntegerType())
VarB = udf(lambda x : x / 638, DoubleType())

result = run_query2(q1, posts_df, posts_df)
result.display()
df_tags = result

df_tags = df_tags.select('*', split("Tags", ">").alias("splits"))
df_tags = df_tags.withColumn("Splits", count_splits(df_tags["splits"]))
df_tags = df_tags.select("OwnerUserId", "Splits")
df_varB = df_tags.withColumn("Splits", VarB(df_tags["splits"]))
df_varB = df_varB.withColumnRenamed("OwnerUserId", "Id")
df_varB = df_varB.filter(df_varB["Id"].isNotNull())
df_varB = df_varB.groupBy("Id").sum("Splits")
df_varB.display()







Id Tags OwnerUserId 9 51 10 22 21 14 23 97 24 14 25 104 26 115 27 108 28 118 29 53 30 26 33 132 37 9 40 104 42 59 43 115 44 59 45 84 46 21 47 62 51 62 53 116 57 84 58 119 60 118 62 26 64 84 70 178 72 22 73 14 74 64 78 178 82 172 83 26 84 26 85 26 87 108 90 156 91 84 92 21 93 21 95 9 96 115 97 158 101 92 106 34 111 26 112 70 120 92 121 92 122 180 126 26 131 26 132 172 133 26 135 227 139 118 140 118 153 84 154 229 156 118 157 36 158 84 160 156 161 21 162 59 163 229 164 208 165 <.net> 108 166 21 170 84 171 172 173 158 174 172 179 97 182 21 183 262 185 108 187 84 188 82 190 173 193 178 194 108 195 92 197 108 198 88 200 278 201 97 202 108 203 265 204 250 206 172 207 21 208 26 210 283 212 21 213 59 214 295 216 108 217 82 219 178 220 250 221 250 222 43 226 290 229 21 230 178 232 375 233 381 236 381 237 59 238 395 241 403 242 406 243 432 244 434 245 432 246 434 247 434 248 434 249 434 250 434 251 451 254 463 255 466 256 434 257 11 258 434 259 454 260 454 261 480 263 478 264 490 268 418 269 97 270 478 271 418 272 478 273 498 274 434 275 508 276 434 277 518 278 524 279 514 282 514 284 84 285 553 287 108 288 381 290 571 291 381 292 587 293 26 294 602 295 386 296 14 297 418 298 609 300 609 301 609 302 434 303 609 306 434 308 434 309 638 311 434 312 655 314 178 316 514 317 680 318 471 320 574 325 434 327 115 328 456 330 729 331 108 332 108 335 478 336 733 337 735 338 737 339 514 340 471 341 743 343 754 345 554 346 554 347 733 348 456 349 780 351 735 355 434 357 797 358 587 359 743 360 801 362 737 363 84 364 158 365 386 366 780 368 791 372 381 373 434 374 780 377 434 379 842 380 842 381 842 386 780 388 780 389 515 390 21 391 735 393 478 394 548 395 548 396 548 397 733 399 735 400 846 401 791 402 387 404 548 405 548 407 227 408 178 409 889 413 890 414 548 415 548 416 381 417 414 419 901 420 178 425 791 426 122 428 906 429 846 431 743 432 241 433 241 434 515 436 548 439 780 440 922 442 548 443 474 444 418 445 609 446 322 448 418 449 21 450 941 452 403 456 322 457 548 459 178 460 984 462 548 463 984 464 984 465 984 469 906 470 941 471 944 475 418 478 null 480 1004 481 59 482 178 483 968 484 1011 485 1015 487 1011 489 1011 491 941 493 1029 495 1011 496 434 498 819 499 819 500 514 501 119 502 108 503 1057 506 1011 507 780 508 984 509 1015 510 941 512 609 513 964 514 609 517 119 519 454 520 1085 521 1061 522 1011 523 1096 525 406 526 1103 529 159 532 418 533 1085 534 883 535 941 539 1127 541 322 543 793 544 793 545 1011 546 1011 549 1011 550 418 552 941 555 434 556 434 560 1085 561 964 562 1155 563 1155 567 1163 568 1155 569 1164 570 434 572 609 573 1176 574 1155 575 1155 576 1155 577 743 580 1155 581 1155 582 743 583 924 584 924 585 924 588 97 591 791 592 172 593 172 597 478 598 1155 599 1155 602 984 603 434 604 434 606 1220 609 1224 611 941 612 1085 614 968 615 1237 617 97 618 1193 619 21 620 953 621 553 622 1085 624 454 625 816 626 1155 627 843 628 843 629 62 630 609 635 21 637 1085 642 159 643 127 645 178 648 984 649 984 650 172 651 172 652 172 654 964 656 1301 659 1289 663 375 665 1111 666 786 667 984 668 1329 669 548 672 609 675 1279 676 1355 680 1279 682 1360 684 1367 686 574 687 548 688 1367 689 548 690 1301 691 1301 692 1301 693 1301 695 684 698 1379 699 1384 702 1367 703 1390 705 1399 707 684 708 84 709 1131 712 984 714 2433 715 434 718 684 719 2443 720 434 721 548 722 2452 723 127 724 172 725 172 727 2459 728 2452 729 375 732 1279 733 471 734 21 735 1367 737 2485 740 1350 741 381 742 2452 743 791 746 2452 747 791 751 941 752 2452 753 2489 754 2513 755 984 756 2515 757 325 759 2452 763 1329 764 802 766 159 767 2544 768 null 769 null 770 null 771 null 773 154 774 2452 775 2575 776 2583 779 2600 780 2452 781 548 782 2452 785 941 788 434 789 802 790 381 792 743 794 2648 795 375 796 2433 797 548 798 381 799 381 800 381 801 2433 803 381 805 2452 807 548 809 2452 811 2513 813 471 814 2452 816 2668 817 null 818 941 819 471 821 471 822 178 824 1279 826 325 827 548 828 192 830 471 833 325 835 2711 836 2452 837 471 839 63174 841 381 843 2452 845 

Id sum(Splits) 463 0.004702194357366771 471 0.19592476489028204 11141 0.031347962382445145 13285 0.032915360501567396 14450 0.009404388714733543 15790 0.017241379310344827 9465 0.023510971786833854 27760 0.012539184952978054 29054 0.006269592476489028 29744 0.003134796238244514 45615 0.0219435736677116 46465 0.004702194357366771 50223 0.007836990595611285 51595 0.001567398119122257 57178 0.017241379310344827 57380 0.007836990595611285 49717 0.007836990595611285 70863 0.001567398119122257 57693 0.007836990595611285 74852 0.006269592476489028 80579 0.004702194357366771 85321 0.006269592476489028 43527 0.006269592476489028 24663 0.003134796238244514 90817 0.004702194357366771 91299 0.007836990595611285 92080 0.003134796238244514 31951 0.006269592476489028 69478 0.21630094043887144 31528 0.006269592476489028 89056 0.007836990595611285 111515 0.001567398119122257 108560 0.006269592476489028 109608 0.006269592476489028 737 0.006269592476489028 1127 0.014106583072100312 4929 0.003134796238244514 9182 0.004702194357366771 10462 0.004702194357366771 10914 0.001567398119122257 12626 0.003134796238244514 18282 0.009404388714733543 39513 0.032915360501567396 43270 0.003134796238244514 43921 0.007836990595611285 43256 0.006269592476489028 16916 0.003134796238244514 49699 0.006269592476489028 35654 0.003134796238244514 57771 0.007836990595611285 58196 0.03918495297805642 40542 0.004702194357366771 63464 0.036050156739811906 67388 0.001567398119122257 68575 0.007836990595611285 69743 0.006269592476489028 14148 0.007836990595611285 78700 0.006269592476489028 85858 0.006269592476489028 86074 0.003134796238244514 107339 0.01567398119122257 38707 0.007836990595611285 113402 0.009404388714733543 2711 0.004702194357366771 3213 0.001567398119122257 8257 0.003134796238244514 9383 0.003134796238244514 13009 0.003134796238244514 13875 0.004702194357366771 15173 0.003134796238244514 16680 0.004702194357366771 17809 0.004702194357366771 18502 0.003134796238244514 23144 0.02664576802507837 28581 0.004702194357366771 42613 0.001567398119122257 37657 0.003134796238244514 38582 0.006269592476489028 42926 0.0438871473354232 60001 0.006269592476489028 60068 0.007836990595611285 75102 0.006269592476489028 76236 0.001567398119122257 67068 0.1018808777429467 80906 0.006269592476489028 29109 0.006269592476489028 90674 0.007836990595611285 96978 0.053291536050156726 98521 0.007836990595611285 63469 0.004702194357366771 101552 0.004702194357366771 105824 0.006269592476489028 87028 0.025078369905956112 107215 0.004702194357366771 108159 0.003134796238244514 94792 0.007836990595611285 109859 0.007836990595611285 110203 0.004702194357366771 46818 0.004702194357366771 451 0.004702194357366771 2443 0.001567398119122257 6597 0.003134796238244514 8650 0.012539184952978056 9946 0.006269592476489028 17712 0.001567398119122257 19161 0.2053291536050157 15604 0.009404388714733543 23367 0.003134796238244514 31275 0.006269592476489028 32554 0.003134796238244514 35713 0.006269592476489028 3089 0.003134796238244514 44623 0.001567398119122257 49696 0.004702194357366771 56490 0.003134796238244514 59229 0.012539184952978056 62400 0.001567398119122257 62606 0.003134796238244514 30802 0.006269592476489028 68386 0.004702194357366771 70109 0.003134796238244514 76202 0.036050156739811906 78539 0.007836990595611285 82049 0.003134796238244514 43108 0.034482758620689655 83613 0.003134796238244514 77373 0.007836990595611285 87349 0.004702194357366771 67574 0.006269592476489028 87742 0.003134796238244514 84602 0.029780564263322887 91919 0.006269592476489028 67050 0.07366771159874609 98306 0.009404388714733543 83663 0.009404388714733541 109250 0.004702194357366771 78269 0.003134796238244514 110876 0.003134796238244514 107002 0.004702194357366771 883 0.003134796238244514 9597 0.010971786833855798 14779 0.05642633228840126 18328 0.004702194357366771 16117 0.003134796238244514 19067 0.006269592476489028 8501 0.04388714733542319 25742 0.02037617554858934 10815 0.004702194357366771 29714 0.015673981

In [0]:
df_varA
df_varB
merged_df = df_varA.join(df_varB, on=["Id"])
merged_df.display()

correlation = merged_df.corr("Reputation", "sum(Splits)", "pearson")
print(correlation)

Id Reputation sum(Splits) 463 71 0.004702194357366771 471 1842 0.19592476489028204 11141 364 0.031347962382445145 13285 188 0.032915360501567396 14450 131 0.009404388714733543 15790 139 0.017241379310344827 9465 1938 0.023510971786833854 27760 11 0.012539184952978054 29054 96 0.006269592476489028 29744 26 0.003134796238244514 45615 251 0.0219435736677116 46465 111 0.004702194357366771 50223 21 0.007836990595611285 51595 1 0.001567398119122257 57178 126 0.017241379310344827 57380 121 0.007836990595611285 49717 11 0.007836990595611285 70863 286 0.001567398119122257 57693 51 0.007836990595611285 74852 1 0.006269592476489028 80579 11 0.004702194357366771 85321 16 0.006269592476489028 43527 101 0.006269592476489028 24663 154 0.003134796238244514 90817 11 0.004702194357366771 91299 1 0.007836990595611285 92080 21 0.003134796238244514 31951 101 0.006269592476489028 69478 910 0.21630094043887144 31528 121 0.006269592476489028 89056 1 0.007836990595611285 737 146 0.006269592476489028 1127 574 0.014106583072100312 4929 11 0.003134796238244514 9182 111 0.004702194357366771 10462 1473 0.004702194357366771 10914 1 0.001567398119122257 12626 151 0.003134796238244514 18282 176 0.009404388714733543 39513 338 0.032915360501567396 43270 131 0.003134796238244514 43921 793 0.007836990595611285 43256 126 0.006269592476489028 16916 1 0.003134796238244514 49699 36 0.006269592476489028 35654 11 0.003134796238244514 57771 121 0.007836990595611285 58196 149 0.03918495297805642 40542 16 0.004702194357366771 63464 39 0.036050156739811906 67388 1 0.001567398119122257 68575 101 0.007836990595611285 69743 111 0.006269592476489028 14148 101 0.007836990595611285 78700 1 0.006269592476489028 85858 1 0.006269592476489028 86074 11 0.003134796238244514 107339 11 0.01567398119122257 38707 101 0.007836990595611285 2711 157 0.004702194357366771 3213 141 0.001567398119122257 8257 11 0.003134796238244514 9383 101 0.003134796238244514 13009 1 0.003134796238244514 13875 41 0.004702194357366771 15173 21 0.003134796238244514 16680 175 0.004702194357366771 17809 36 0.004702194357366771 18502 11 0.003134796238244514 23144 186 0.02664576802507837 28581 123 0.004702194357366771 42613 109 0.001567398119122257 37657 103 0.003134796238244514 38582 111 0.006269592476489028 42926 271 0.0438871473354232 60001 11 0.006269592476489028 60068 1 0.007836990595611285 75102 21 0.006269592476489028 76236 21 0.001567398119122257 67068 139 0.1018808777429467 80906 116 0.006269592476489028 29109 293 0.006269592476489028 90674 1 0.007836990595611285 96978 91 0.053291536050156726 98521 1 0.007836990595611285 63469 111 0.004702194357366771 101552 1 0.004702194357366771 105824 21 0.006269592476489028 87028 60 0.025078369905956112 107215 13 0.004702194357366771 108159 1 0.003134796238244514 94792 103 0.007836990595611285 46818 101 0.004702194357366771 451 156 0.004702194357366771 2443 141 0.001567398119122257 6597 111 0.003134796238244514 8650 201 0.012539184952978056 9946 146 0.006269592476489028 17712 21 0.001567398119122257 19161 1319 0.2053291536050157 15604 241 0.009404388714733543 23367 111 0.003134796238244514 31275 111 0.006269592476489028 32554 131 0.003134796238244514 35713 19 0.006269592476489028 3089 111 0.003134796238244514 44623 111 0.001567398119122257 49696 111 0.004702194357366771 56490 105 0.003134796238244514 59229 356 0.012539184952978056 62400 11 0.001567398119122257 62606 11 0.003134796238244514 30802 11 0.006269592476489028 68386 191 0.004702194357366771 70109 11 0.003134796238244514 76202 167 0.036050156739811906 78539 1 0.007836990595611285 82049 11 0.003134796238244514 43108 165 0.034482758620689655 83613 36 0.003134796238244514 77373 121 0.007836990595611285 87349 11 0.004702194357366771 67574 11 0.006269592476489028 87742 41 0.003134796238244514 84602 243 0.029780564263322887 91919 11 0.006269592476489028 67050 305 0.07366771159874609 98306 136 0.009404388714733543 83663 141 0.009404388714733541 78269 185 0.003134796238244514 107002 1 0.004702194357366771 883 123 0.0031

0.7724926022809913



I first selected the reputation of each user with their user_id. Then I joined the Posts to find all the Tags for answers by Users. So I selected all the tags for each question and for each user. I calculated the amount of tags in one question divided by the amount of tags in total (638). Then I grouped all the users together and added the user interest diversity. Then I merged VarA and VarB on their respective user_ids. Then I had a table with Id, Reputation and their user interest diversity. Then I calculated the pearson coefficient on their reputation and their user interest diversity. We found a pearson coefficient of 0.77249 which means there is a high corellation between user interest diversity and their reputation. This means our users with high reputation have a more general interest then specific interests.

